merkle-tree/src/merkle_tree.rs

```diff
@@ -139,6 +139,85 @@ impl MerkleTree {
         self.nodes.iter().last()
     }
 
+    fn append_nodes(nodes: &mut [Hash], mut leaf_count: usize, leaves: Vec<Hash>) -> usize {
+        for mut leaf in leaves {
+            let mut layer: usize = 0;
+            leaf_count += 1;
+            let mut cursor = leaf_count;
+            while (cursor & 1) == 0 {
+                let arg1 = &leaf;
+                let arg2 = &nodes[layer];
+                //todo: verify and implement jump's fancy avx implementation
+                leaf = hash_intermediate!(arg2, arg1);
+                layer += 1;
+                cursor >>= 1;
+            }
+
+            nodes[layer] = leaf;
+        }
+        leaf_count
+    }
+
+    fn private_depth(leaf_count: usize) -> usize {
+        if leaf_count <= 1 {
+            leaf_count
+        } else {
+            (63 - (leaf_count - 1).leading_zeros()) as usize + 2
+        }
+    }
+
+    fn commit_finish(nodes: &mut [Hash], leaf_count: usize) -> Hash {
+        let root_idx: usize = Self::private_depth(leaf_count) - 1;
+
+        if !leaf_count.is_power_of_two() {
+            let mut layer = leaf_count.trailing_zeros() as usize;
+            let mut layer_count = leaf_count >> layer;
+            let mut tmp = nodes[layer];
+
+            while layer_count > 1 {
+                tmp = if (layer_count & 1) != 0 {
+                    let arg1 = &tmp;
+                    let arg2 = &tmp;
+                    hash_intermediate!(arg2, arg1)
+                } else {
+                    let arg1 = &tmp;
+                    let arg2 = &nodes[layer];
```

Comment:
again, these `arg1`, `arg2` are redundant and confusing.

Reply:
See above